Adding yumrepo module

jtyr opened the pull request:
This patch is adding a new module which allows to add and remove YUM repository definitions. The module implements all repository options as described in the `yum.conf` manual page.

resmo commented on packaging/os/yumrepo.py:
```diff
@@ -0,0 +1,501 @@
+DOCUMENTATION = '''
+---
+module: yumrepo
+version_added: 2.0
```
please add quotes to the 2.0: `version_added: "2.0"`

jtyr replied:
Fixed

resmo commented on packaging/os/yumrepo.py:
```diff
@@ -0,0 +1,501 @@
+    C(YUMREPO_DIR).
+  - The repo file will be automatically deleted if it contains no repository.
+
+author: Jiri Tyr
```
add you github handle as well. e.g. `author: "Jiri Tyr (@jtyr)"`

jtyr replied:
Fixed

robynbergeron commented:
Hi @jtyr -- Thanks for submitting this new module to Ansible Extras! This module is now in community review, a process that is open to all Ansible users. In order for this module to be approved, it must gain the following votes:

“works_for_me”: If you have tested the module thoroughly, including testing of all of the module’s options, and if the module works for you, please add “works_for_me” in the comments.

“passes_guidelines”: If you have gone through the module guidelines and the module meets all of the requirements, please add “passes_guidelines” in the comments. Guidelines are available here: http://docs.ansible.com/developing_modules.html#module-checklist

“needs_revision”: If the module fails to work for you, or if it doesn’t meet guidelines, please add “needs_revision” in the comments with details about what needs to be fixed.

When a module has both “works_for_me” and “passes_guidelines” tags, we will promote the module for inclusion in Ansible Extras.  At this point, you will be expected to maintain the module by fixing bugs and evaluating pull requests in a timely manner. 

Thanks again for submitting your Ansible module!

robynbergeron commented:
(Also: thanks @resmo for comments thus far!)

jtyr commented:
@robynbergeron Thanks for the information. Please could you advise me what's the best way how to get somebody to review and tag the module with the “works_for_me” and “passes_guidelines” tags? Should I contact somebody from the [Subject Matter Experts](https://github.com/ansible/ansible-modules-extras/blob/devel/REVIEWERS.md#subject-matter-experts) or should I just wait till somebody is willing to review it?

robynbergeron commented:
Hi @jtyr -- 

You can read a bit more about the above referenced process here -- it is freshly introduced. ;) http://robyn.io/2015/09/28/ansible-extras-and-you/

We don't have a specific "best way" just yet -- and most of the subject matter expert folks that we have are more specific to areas other than this.  My suggestion would be to give it a little shout-out on twitter, facebook, or some other social media type place -- or maybe review someone else's outstanding new module PR and hope that the karma is returned :) We're hoping that widening the pool of potential reviewers will help out, but it may take a little bit of time for people to know that they really can help out :) 

I will give it a little bump on twitter and a shoutout to my Fedora friends and see if maybe anyone is willing to help out.

jtyr commented:
@robynbergeron Thanks for the explanation. I would definitely like to help to review some of the pull requests but first I would like to know if the one I have written satisfies the standard in order I can use the same measure for the others. I really appreciate your help and I hope I will be able to pay it back soon.

mgedmin commented on packaging/os/yumrepo.py:
```diff
@@ -0,0 +1,501 @@
+
+
+from ansible.constants import get_config
+from ansible.constants import load_config_file
```
I don't think you can use those in modules.

The way Ansible modules work: your module's source code is pre-processed, treating `from ansible.module_utils.SOMETHING import *` as an include directive, then some placeholders in those includes get replaced with module arguments etc., then the preprocessed source is SCPed over to the remote and executed as a single file.

There's no requirement for Ansible to be installed on remote hosts, so you can't rely on ansible imports succeeding, with the exception of those special include directives that masquerade as Python imports.

jtyr replied:
Thanks for the explanation. I have removed this kind of configuration and introduced a new module parameter instead.

mgedmin commented on packaging/os/yumrepo.py:
```diff
@@ -0,0 +1,501 @@
+from ansible.constants import get_config
+from ansible.constants import load_config_file
+from ansible.module_utils.basic import *
+from ansible.module_utils.urls import *
```
These should appear at the bottom.  One reason for that is to maintain useful line number information in tracebacks (see the above comment for a description of the textual preprocessing taking place; if the contents of those module utils get included here, all the line numbers below will shift).

This is mentioned in the [Ansible module checklist](http://docs.ansible.com/ansible/developing_modules.html#module-checklist) BTW.

jtyr replied:
I have moved these two imports right above the `main()` call.

mgedmin commented on packaging/os/yumrepo.py:
```diff
@@ -0,0 +1,501 @@
+  failovermethod:
+    required: false
+    choices: [roundrobin, priority]
+    default: roundrobin
```
Stylstic issue: choices/default are shown as quoted strings elsewhere, but not here.

jtyr replied:
I have fixed the quotation to be more consistent. I'm quoting only the values which might be converted to boolean (e.g. yes, no).

mgedmin commented on packaging/os/yumrepo.py:
```diff
@@ -0,0 +1,501 @@
+    required: false
+    default: null
+    description:
+      - .
```
This seems incomplete.

jtyr replied:
I have added the description.

mgedmin commented on packaging/os/yumrepo.py:
```diff
@@ -0,0 +1,501 @@
+            value = self.params[key]
+
+            # Convert boolean value to integer
+            if type(value).__name__ == 'bool':
```
Please use `if isinstance(value, bool):`

jtyr replied:
Fixed.

mgedmin commented on packaging/os/yumrepo.py:
```diff
@@ -0,0 +1,501 @@
+
+            repo_string += "\n"
+
+        return repo_string
```
It would be simpler to 

``` python
from StringIO import StringIO
...
    def dump(self):
        buf = StringIO()
        self.repofile.write(buf)
        return buf.getvalue()
```

mgedmin replied:
On second thought I'm not sure that would sort the sections the right way, especially if you remove and add back the same section.  Hm.

On the other hand, you're not sorting the keys.

jtyr replied:
I have added the key sorting.

larsks commented:
This playbook is incorrect w/r/t to what the module requires:

```
- hosts: localhost
  tasks:
    - yumrepo:
        name: testrepo
        baseurl: http://localhost/packages/
        exclude:
          - broken-package
          - anotherpackage
```

The above should either (a) work, by translating the YAML list into
the correct syntax for a yum repo configuration, or (b) it should
raise an error.

As it stands, it results in:

```
[testrepo]
baseurl = http://localhost/packages/
exclude = ['broken-package', 'anotherpackage']
```

I would vote for (a), so that it works naturally with YAML lists.

larsks commented:
Should this module also accept `file`-like parameters for setting group/owner/etc?

larsks commented on packaging/os/yumrepo.py:
```diff
@@ -0,0 +1,502 @@
+        expire.
+      - Default value is 6 hours.
+  name:
+    required: true
```
A minor request, but maybe place any `required` options at the top of this list so they're easier to spot?

jtyr replied:
According to the documentation of other (core) modules (e.g. group, template) the parameters are sorted alphabetically.

larsks replied:
That's fine.  Like I said, minor, and I figured it was probably canonical.

bcoca replied:
parameters get sorted on display, so in the file do whatever is more readable for the maintainers

jtyr commented:
@larsks Thanks for the review. The `exclude` list should be "space separated list" as written in the documentation. I'm not sure if there is any other Ansile module which would allow to use data structure as the value of a parameter.

larsks commented:
So you should go with choice (b) then, which was "raise an error", rather than string-ifying the list.

jtyr commented:
@larsks Good point with the file-like parameters!

larsks commented:
For what it's worth, there _are_ ansible modules that accept data structures as parameters (e.g., cloud/openstack/os_server.py).

jtyr commented:
@larsks I have added the support for all file params. I have also added the support for the YAML array in  the case of the `exclude` and the `includepkgs` parameter. Please review and comment with “works_for_me” and/or “passes_guidelines” if you think it works well.

larsks commented:
works_for_me

With the most recent changes, the following playbook does "what I expect":

```
- hosts: localhost
  tasks:
    - yumrepo:
        name: testrepo
        baseurl: http://localhost/packages/
        exclude:
          - broken-package
          - anotherpackage
        includepkgs: thisone tahtone
        gpgcheck: false
        owner: root
        group: wheel
```

robynbergeron commented:
Hi @jtyr -- thanks for your updates to this PR and your work with @larsks and @mgedmin on this. (Also, HI @larsks! Nice to see you here, used to seeing you in Fedora / RDO land :D)

It looks like we now have a works_for_me vote -- which means we now just need a passes_guidelines vote. Perhaps @larsks or @mgedmin are willing to do that -- or anyone else can as well.

Friendly reminder for anyone who may want to review this for guidelines adherence:

“passes_guidelines”: If you have gone through the module guidelines and the module meets all of the requirements, please add “passes_guidelines” in the comments. Guidelines are available here: http://docs.ansible.com/developing_modules.html#module-checklist

“needs_revision”: If the module fails to work for you, or if it doesn’t meet guidelines, please add “needs_revision” in the comments with details about what needs to be fixed.

robynbergeron commented:
Also: More out of curiosity -- I am wondering if / how easily someone could use this as a model to do a similar module for dnf, to go along with dnf.py in extras. Thoughts? @abadger -- maybe you have some thoughts on this as well? It would be nice just from a parity standpoint for some those folks who make the move from yum to dnf on their systems at some point in the future. (I can totally file a feature request / idea issue for this, but figured I'd just ask the question to a captive audience -- answers to this question either way definitely won't block the inclusion of this module once it has the required votes :D)

larsks commented:
@robynbergeron This module should work fine with dnf as is.  DNF currently uses the same repository configuration (in the same location) as yum.

larsks commented:
needs_revision: Pretty much everything looks fine, except that I don't see documentation for:
- Return: document the return structure of the module

jtyr commented:
@larsks This module returns only the standard structure (`changed`, `name`, `state`) which I think is not worth documenting.

larsks commented:
That's fine.  This is my first time reviewing a module, and I don't have a sense how strict the community is w/r/t to those requirements.  I'll wait a bit, but if nobody else has comments I'll just declare it good by fiat later today :)

mgedmin commented on packaging/os/yumrepo.py:
```diff
@@ -0,0 +1,529 @@
+
+        # Set options
+        for key in sorted(self.params.keys()):
+            value = self.params[key]
```
Minor: you could simplify these two lines to

``` python
for key, value in sorted(self.params.items()):
```

jtyr replied:
Fixed.

mgedmin commented on packaging/os/yumrepo.py:
```diff
@@ -0,0 +1,529 @@
+
+    # Save the file only if not in check mode and if there was a change
+    if not module.check_mode and changed:
+        yumrepo.save()
```
This could raise (because of file permissions/missing directories/out of disk space/whatever).  AFAIU the [module guidelines](http://docs.ansible.com/ansible/developing_modules.html#module-checklist) ask that you catch the IOError here and replace it with a `module.fail_json(...)` call.

jtyr replied:
I have added the exception handling into the `save()` method.

mgedmin commented:
LGTM, except for the missing error handling.

(Grain of salt: I am not an experienced Ansible module reviewer either.)

mgedmin commented on packaging/os/yumrepo.py:
```diff
@@ -0,0 +1,529 @@
+            repo_file = self.params['file']
+
+        # Set dest; also used to set dest parameter for the FS attributes
+        self.params['dest'] = "%s/%s.repo" % (repos_dir, repo_file)
```
Why not `os.path.join(repos_dir, repo_file)`?  Would avoid double slashes if the user specifies `repos_dir=/with/a/trailing/slash/`.

jtyr replied:
Fixed.

larsks commented:
Yay, another reviewer! I like @mgedmin's suggestions, particularly w/r/t catching ioerror exceptions raised by file operations.

jtyr commented:
@mgedmin Thank you very much for your comments. I have fixed all the issues in the latest update (see the diff bellow). Please review the changes and and let me know if it looks good to you.

@larsks @mgedmin I have replaced the `hidden_params` variable with the `allowed_params` variable which helps to make sure that only params from this list will be in the final repo file (see the diff bellow). Please test the module and comment with "works_for_me" and/or "passes_guidelines" if there are no further problems.

``` diff
diff --git a/packaging/os/yumrepo.py b/packaging/os/yumrepo.py
index cb14a90..13dbaad 100644
--- a/packaging/os/yumrepo.py
+++ b/packaging/os/yumrepo.py
@@ -336,16 +336,22 @@ class YumRepo(object):
     section = None
     repofile = ConfigParser.RawConfigParser()

+    # List of parameters which will be allowed in the repo file output
+    allowed_params = [
+        'bandwidth', 'baseurl', 'cost', 'enabled', 'enablegroups', 'exclude',
+        'failovermethod', 'gpgcakey', 'gpgcheck', 'gpgkey', 'http_caching',
+        'includepkgs', 'keepalive', 'metadata_expire', 'metalink',
+        'mirrorlist', 'mirrorlist_expire', 'name', 'password', 'protect',
+        'proxy', 'proxy_password', 'proxy_username', 'repo_gpgcheck',
+        'retries', 'skip_if_unavailable', 'sslcacert',
+        'ssl_check_cert_permissions', 'sslclientcert', 'sslclientkey',
+        'sslverify', 'throttle', 'timeout', 'username']
+
     # List of parameters which can be a list
     list_params = ['exclude', 'includepkgs']

-    # List of keys which will be excluded from the final output
-    hidden_params = [
-        'file', 'repoid', 'reposdir', 'state', 'dest', 'checkmode',
-        'mode', 'owner', 'group', 'seuser', 'serole', 'setype', 'selevel']
-
     def __init__(self, module):
-        # To be albe to use fail_json
+        # To be able to use fail_json
         self.module = module
         # Shortcut for the params
         self.params = self.module.params
@@ -364,7 +370,7 @@ class YumRepo(object):
             repo_file = self.params['file']

         # Set dest; also used to set dest parameter for the FS attributes
-        self.params['dest'] = "%s/%s.repo" % (repos_dir, repo_file)
+        self.params['dest'] = os.path.join(repos_dir, "%s.repo" % repo_file)

         # Read the repo file if it exists
         if os.path.isfile(self.params['dest']):
@@ -387,9 +393,7 @@ class YumRepo(object):
                 'adding a new repo.')

         # Set options
-        for key in sorted(self.params.keys()):
-            value = self.params[key]
-
+        for key, value in sorted(self.params.items()):
             if key in self.list_params and isinstance(value, list):
                 # Join items into one string for specific parameters
                 value = ' '.join(value)
@@ -398,18 +402,36 @@ class YumRepo(object):
                 value = int(value)

             # Set the value only if it was defined (default is None)
-            if value is not None and key not in self.hidden_params:
+            if value is not None and key in self.allowed_params:
                 self.repofile.set(self.section, key, value)

     def save(self):
         if len(self.repofile.sections()):
             # Write data into the file
-            fd = open(self.params['dest'], 'wb')
-            self.repofile.write(fd)
-            fd.close()
+            try:
+                fd = open(self.params['dest'], 'wb')
+            except IOError:
+                self.module.fail_json(
+                    msg='Can not open repo file %s.' %
+                    self.params['dest'])
+
+            try:
+                try:
+                    self.repofile.write(fd)
+                except Error:
+                    self.module.fail_json(
+                        msg='Can not write repo file %s.' %
+                        self.params['dest'])
+            finally:
+                fd.close()
         else:
             # Remove the file if there are not repos
-            os.remove(self.params['dest'])
+            try:
+                os.remove(self.params['dest'])
+            except OSError:
+                self.module.fail_json(
+                    msg='Can not remove empty repo file %s.' %
+                    self.params['dest'])

     def remove(self):
         # Remove section if exists
@@ -424,11 +443,7 @@ class YumRepo(object):
             repo_string += "[%s]\n" % section

             for key, value in sorted(self.repofile.items(section)):
-                # This is here to prevent the checkmod param to appear in the
-                # dump output (Dunno how it appears here as it's filtered out
-                # in the add() method. Please help!).
-                if key not in self.hidden_params:
-                    repo_string += "%s = %s\n" % (key, value)
+                repo_string += "%s = %s\n" % (key, value)

             repo_string += "\n"

```

mgedmin commented on packaging/os/yumrepo.py:
```diff
@@ -0,0 +1,547 @@
+                fd = open(self.params['dest'], 'wb')
+            except IOError:
+                self.module.fail_json(
+                    msg='Can not open repo file %s.' %
```
Typo: should be "Cannot" (or "Can't").

jtyr replied:
Fixed.

mgedmin commented on packaging/os/yumrepo.py:
```diff
@@ -0,0 +1,547 @@
+                    self.repofile.write(fd)
+                except Error:
+                    self.module.fail_json(
+                        msg='Can not write repo file %s.' %
```
Same here.

jtyr replied:
Fixed.

mgedmin commented on packaging/os/yumrepo.py:
```diff
@@ -0,0 +1,547 @@
+                os.remove(self.params['dest'])
+            except OSError:
+                self.module.fail_json(
+                    msg='Can not remove empty repo file %s.' %
```
Same here.

jtyr replied:
Fixed.

robynbergeron commented:
Hi @jtyr -- thanks for your quick revisions! :)

@mgedmin & @larsks  -- if you can take a look at this and let us know how these revisions work, that would be awesome.

I'd also like to take a moment to add that even though both of you are not "experienced" module reviewers, you're doing great and doing exactly what we believed community members would do when we moved to the new module review guidelines for extras: make reviews for quality, in good faith. 
https://github.com/ansible/ansible-modules-extras/blob/devel/REVIEWERS.md

You're both clearly paying attention to the guidelines, and making sure it works and is a good module, and that kind of care is largely what ensures that things that go in will be good :) 

Just to be clear about where we are at this point: we need a passes_guidelines vote, and a works_for_me vote; @larsks, you gave a works_for_me vote previously, but I suspect the module has changed enough since that point that giving it another test for functionality would be a good idea.

Thanks, everyone!

mgedmin commented:
passes_guidelines

I haven't actually gone through all of them _this time_ -- but last time I did, and the missing error handling was the only thing that stood out.  Well, error handling and the lack of description of the module return value, [mentioned earlier](https://github.com/ansible/ansible-modules-extras/pull/1016#issuecomment-147401562).

larsks commented:
passes_guidelines
works_for_me

The module continues to work as expected, and now provides a better error message if file permissions prevent an operation from succeeding.

jtyr commented:
@robynbergeron What about to celebrate [the great news](http://www.ansible.com/blog/red-hat) by merging this great module? ;o)

jtyr commented:
@robynbergeron @gregdek Any chance this PR gets some attention any time soon?

gregdek commented:
Thanks @jtyr -- marking for inclusion.

jtyr commented:
@robynbergeron @gregdek It has been more than 3 weeks since this PR was marked as "shipit" and it still wasn't merged. Is there any other step which needs to be done before it gets merged?

bcoca commented:
Sorry for the delay, but we had stopped merging new modules until we cut stable-2.0 branch and were concentrating on bugs to launch 2.0.

for 2.1 and above we are requiring return docs to be documented, which this module does not have.

jtyr commented:
@bcoca Thanks for the information. It's pity that the module won't make it into the v2.0 :o( I have added the missing `RETURN` documentation. Please review it.

bcoca commented:
sorry about the exclusion, but we had to cut off somewhere and missed this by a hair. On the bright side 2.1 will not take as long as 2.0, it is just an incremental release and not a core rewrite, I don't think anyone wants to do another any time soon.

jtyr commented:
@bcoca I understand. Would it be possible to cherry-picked this commit and shipped with v2.0.1?
